hera_cal/firstcal.py

```diff
@@ -462,6 +469,106 @@ def firstcal_run(files, opts, history):
+        return data_dict
+
+    def _apply_pi_shift(data_dict, invert_these):
+        """Apply a pi shift to visibilities.
+
+        Args:
+            data_dict (dict): Dictionary of visibilities indexed by bl and pol.
+            invert_these (list): list of antennas to add a pi phase shift too
+        """
+        for ai, aj in data_dict.keys():
+            for pol in data_dict[ai, aj].keys():
+                if (ai, pol[0]) in invert_these or (aj, pol[1]) in invert_these:
+                    data_dict[ai,aj][pol] *= -1 
+
+        return data_dict
+    
+    def _search_and_iterate_firstcal(uv, info, option_parser):
+        '''Searches and iterates over firstcal
+
+            Iteratively run firstcal and look for rotated antennas.
+            If rotated antennas found, fix the antenna and rerun firstcal.
+            Saves rotated antennas to a json file.
+
+        Args:
+            uv (pyuvdata.UVData): UVData object
+            info (FirstcalRedundantInfo): info object
+            option_parser (OptionParser): option parser object
+
+        Return:
+            dict : firstcal solutions
+            list : list of rotated antennas
+        '''
+        switched_history = []  # keep track of all rotated antennas
+        switched = []  # keep track of rotated antennas in each iteration
+        niters = 0
+        try:
+            ex_ants = map(int, option_parser.ex_ants.split(','))
+        except ValueError:
+            ex_ants = []
+
+        while niters == 0 or len(switched) > 0:
+            datapack, wgtpack = UVData_to_dict([uv], ex_ants=ex_ants)
+            datapack = _apply_pi_shift(datapack, switched)
+            wgtpack = {k: {p: np.logical_not(wgtpack[k][p]) for p in wgtpack[k]} for k in wgtpack} 
+
+            fqs = uv.freq_array[0, :] / 1e9
+
+            # gets phase solutions per frequency.
+            fc = FirstCal(datapack, wgtpack, fqs, info)
+            sols = fc.run(finetune=option_parser.finetune,
+                          verbose=option_parser.verbose,
+                          average=option_parser.average, 
+                          window='none')
+
+            # Now we need to check if antennas are flipped
+            medians = {}
+            cal_data = copy.deepcopy(datapack)
+            cal_data = _apply_first_cal(cal_data, sols, fqs*1e9)
+
+            ratio_bls = [(bl1, bl2) for bls in info.get_reds() for b1, bl1 in enumerate(bls) for bl2 in bls[b1 + 1:] if bl1 != bl2]
+            for i, ((a1, a2), (a3, a4)) in enumerate(ratio_bls):
+                # need to get polarizations if run in dual pol. This works because we are looking at reds
+                # so that if one antenna is x, all should be x.
+                if a1 < info.nant:
+                    pol = 'x'
+                else:
+                    pol = 'y'  
+                if (a1, a2) not in cal_data.keys():
```

Comment:
Need to check data conjugation when flipping a1, a2. Alternatively, look into using datacontainer.